### DISCUSSION TITLE
아이템 46 : 스트림에서는 부작용 없는 함수를 사용하라

### DISCUSSION THREAD

**Discussion by jinan159** [Study-2-Effective-Java/Effective-Java#119] [3. 과제] [answered]
# 0\. 들어가며

스트림에서 '부작용 없는 함수'를 사용하라는 말은 무슨 뜻 일까요?  
'부작용' 이라고 번역된 단어는 원서는 'Side effect' 입니다.

> 즉, 이 함수의 인해 '다른곳에 영향을 끼치지 않는 함수'를 스트림에서 사용해야 한다는 뜻이 됩니다.

# 1\. 순수함수

다른곳에 영향을 끼치지 않는 함수를 함수형 프로그래밍에서는 '순수 함수'라고 하는데요.  
이는 조금 바꿔서 말하면 '같은 입력이 주어지면, 항상 같은 결과를 내는 함수'라고 생각할 수 있습니다.

이를 코드로 한번 살펴볼텐데요.  
먼저 순수함수가 아닌 예를 먼저 살펴봅시다.

```java
class Something {
    private static int num = 1;
    public static void main(String[] args) {
        IntStream.range(0,10).forEach(HelloWorld::plusNumber);
        System.out.println(num); // 46
    }

    private static void plusNumber(int i) {
        num += i;
    }
}
```

plusNumber 라는 함수는 주어진 i 를 Something 이라는 클래스의 num 에 더하는 역할을 합니다.  
이 함수는 i가 같은 값이 들어오더라도, num 이 달라지면 다른 결과를 만들어냅니다.  
그래서 이 함수는 순수함수가 아니죠.

그래서 보통 순수함수는 외부 요소를 참조하지 않거나, 참조하더라도 변경시키지 않는 경우가 많습니다.  
만약 변경된 결과가 필요하다면, 새로 생성하여 반환해야 합니다.

# 2\. 짧고 명확하게

책에서는 스트림을 잘 사용하면 표현력, 속도, (상황에 따라)병렬성을 얻을 수 있다고 합니다.  
하지만 스트림을 부작용이 있는 함수 혹은 길고 복잡한 로직과 함께 사용한다면 그 진가를 발휘하지 못한다고 합니다.

```java
import java.util.Map;
import java.util.stream.Stream;

Map<String, Long> freq = new HashMap<>();
try (Stream<String> words = new Scanner(file).tokens()) {
    words.forEach(word -> {
        freq.merge(word.toLowerCase(), 1L, Long::sum)
    }
}
```

위 로직은 두 가지 문제가 있습니다.

1.  forEach 에서 외부(freq)에 영향을 주는 부작용 있는 로직이 포함되어 있음
2.  코드가 간결하지 않습니다.

그래서 책에서는 Stream 에서 제공하는 API 들을 통해 아래와 같이 간결한 코드로 만들었습니다.

```java
import static java.util.stream.Collectors.counting;
import static java.util.stream.Collectors.groupingBy;

import java.util.Map;
import java.util.stream.Stream;

Map<String, Long> freq;
try (Stream<String> words = Stream.of("abcdefg", "hijklmn", "opqrstu", "vwxyz", "opqrstu")) {
    freq = words.collect(groupingBy(String::toLowerCase, counting()));
}
```

하지만 저는 스트림에 익숙하지 않아서, 이 코드를 보고 오히려 직관적으로 이해가 되진 않았는데요.  
그래서 잘 활용하기 위해서는 스트림이 제공하는 다양한 기능들을 미리 파악하고 있어야 '스트림 답게' 사용할 수 있겠구나 생각이 들었습니다.

# 3\. 스트림이 제공하는 API 들

![](https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2Ftvn7b%2FbtrYT95XqBz%2FYVpV4OWGubrkXuUEZEwqKk%2Fimg.png)

- Filter : 스트림의 내용을 일부 필터링 하는데 사용된다.
  - ex) words.filter(w -> w.length() > 5); // 5자 이상인 문자들만 뽑음
- Sorted
  - ex) words.sorted(comparing(String::length)); // 길이가 작은 순으로 정렬
- Map
  - ex) words.map(Word::new); // String 을 Word 클래스로 변환
- Collect
  - ex) words.map(Word::new).collect(toList()); // Word 클래스 리스트로 변환
  
위처럼 다양하게 활용 가능한데요.
이 처럼 스트림을 조작하거나 손쉽게 다른 컬렉션으로 변환하거나, 축소하여 저장할 수 있도록 도와줍니다.
간편하게 변환을 지원하는 컬렉션은 List, Set, Collection 입니다.

- List : Collectors.toList()
- Set : Collectors.toSet()
- Collection : Collectors.toCollection(collectionFactory)

이 외에도 groupingBy 라는 메서드로 Map 으로 변환할 수도 있습니다만 이는 단순 변환이 아닌 '축소'에 해당합니다.

이 대표적인 예가 2번에서 살펴봤던 변경 후 예제 코드인데요.

```java
import static java.util.stream.Collectors.counting;
import static java.util.stream.Collectors.groupingBy;

import java.util.Map;
import java.util.stream.Stream;

Map<String, Long> freq;
try (Stream<String> words = Stream.of("abcdefg", "hijklmn", "opqrstu", "vwxyz", "opqrstu")) {
    freq = words.collect(groupingBy(String::toLowerCase, counting()));
}
```

이를 보면 문자열 스트림은 총 5개의 문자열을 가지고 있습니다.
하지만 groupingBy 를 통해서 단어를 key 로 단어의 중복 개수를 value 로 하는 Map 을 반환합니다.
이렇게 되면 key, value 4쌍으로 축소되며 Map 으로 변환되었죠.

# 4. 결론

이 처럼 Stream 을 이용하면 간결하고 명확한 코드를 빠르게 생산해낼 수 있습니다.
그래서 Stream 이 제공하는 기능들에 익숙해 진다면 생산성이 향상될 것이라 생각됩니다.

하지만 팀원들이 모두 익숙한 상황이 아닌 상태라면 협업 관점에서는 좋은 영향이 있을지는 조금 더 고민해봐야할 것 같습니다.

**Comment by Irisation23** [comment]
@jinan159 님께서 말씀해주신 `부작용 없는 함수`라는 개념이 과연 내포하고 있는 의미가 무엇인지에 대해 생각 해 본적이 있습니다.

불변이 아닌 변수는 예상치 못한 상황에 예상치 못한 값으로 변경될 가능성이 있고, 함수형 프로그램은 대입문이 없어서 기본적으로 한번 값이 변수에 할당되고 나면 이후에 값이 변경되지 않는 특징을 가지고 있는것으로 알고 있습니다.

부수효과(Side Effect)를 발생시키지 않기 때문에 표현 식을 언제 어느 때 실행 하더라도 문제가 발생하지 않으며 이는 프로그래머가 신경 쓸 부분이 줄어들게 만들어줍니다. 이러한 특성을 `참조 투명성`이라고 하는데
바로 `참조 투명성`을 지니고 있기 때문에 멀티코어 프로세스에서 `DeadLock`이 빠지지 않는 장점이 생깁니다.

@jinan159 님이 정리해주신 위의 `순수함수` 파트를 잘 소화시키기 위한 내용은 함수형 프로그램 특징과 맞닿아 있었다고 생각합니다.

감사합니다.

> 출처 내용
> > [Java 8에서 왜 함수형 프로그래밍이 도입되었을까?](https://tecoble.techcourse.co.kr/post/2021-09-30-java8-functional-programming/)
> > [부수효과와 참조 투명성](https://jinwooe.wordpress.com/2017/12/21/%EB%B6%80%EC%88%98-%ED%9A%A8%EA%B3%BC-side-effect-%EC%B0%B8%EC%A1%B0-%ED%88%AC%EB%AA%85%EC%84%B1-referential-transparency/)

**Comment by chikeem90** [comment]
스트림,, 이제는 스트림 없이 어떻게 개발했는지는 상상 못 할 정도로,, 저도 적극 사용하고 있습니다.
저도 처음에는 익숙하지 않으니 가독성이 오히려 떨어진다고 느끼기도 했는데,, 코드량 자체가 상당히 간결해지니까 사용하는 것이 훨씬 좋은 거 같습니다. 
정리해주신 것처럼 우선 스트림에서 제공하는 주요한 api들부터 익혀가시면 좋을 거 같아요 ㅎㅎ
filter, map, flatMap, forEach, collect, groupBy 등등 ,, 정말 많이 사용하게 되는거 같아요.

**Comment by JoisFe** [comment, marked as answer]
이전에 #106 디스커션 댓글에서 언급한 순수함수가 여기서 나오는 군요

```
스트림을 이용한 병렬 처리 등을 통해 언급하신 비동기, 논블로킹 방식을 구현할 수 있다는 점 또한 있습니다.
함수형 프로그래밍의 특징은 일반적으로 우리가 생각하는 함수와 다르게

순수 함수가 되어야 한다고 합니다.
함수에서 외부의 상태값을 참조하거나 외부의 상태를 변경해서는 안되는 것입니다.
```

@coalong 님 께서 작성하신 #121 에서 언급하신  
```이처럼 자바로 동시성 프로그램을 작성하기 쉬워지고는 있지만, 안정성(safety)과 응답 가능(liveness) 상태를 유지하기 위해 코드를 올바르게 작성하는 것은 어려운 작업이다.```
과 관련된 것 같으니 같이 보면 좋을 것 같아요
